File: nomorebuild/problems/28.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */

// 해당 문제는 indexOf를 생각하면 편함.

const strStr = (haystack, needle) => {
  // needle이 없거나 길이가 0일때 0을 반환
  // 본문: we will return 0 when needle is an empty string.
  if (!needle.length) return 0;

  // haystack만큼 i회 반복
  for (let i = 0; i < haystack.length; i++) {
    // haystack의 문자열에서 i와 i + needle.length 사이 만큼 substring
    // 여기서 substring시 he, ll, o 으로 출력 되는데 이 값과 needle의 값을 비교하면 됨.
    if (haystack.substring(i, i + needle.length) === needle) {
      return i;
    }
  }

  // haystack의 검색결과가 없을때 -1 반환
  // 본문: Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
  return -1;
};

// 가장 간단하면서 광탈하기 좋은 방법...
// const strStr = (haystack, needle) => {
//     return haystack.indexOf(needle)
// }

File: bugoverdose/28.py
# Runtime: 32 ms, faster than 67.22% of Python3 online submissions for Implement strStr().
# Memory Usage: 14.2 MB, less than 92.16% of Python3 online submissions for Implement strStr().
class Solution:
    def strStr(_, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0
        if len(haystack) >= len(needle):
            for i in range(len(haystack) - len(needle) + 1):  
                if haystack[i] == needle[0]: 
                    for j in range(len(needle)): 
                        if haystack[i + j] != needle[j]:
                            break
                        if j == len(needle) - 1:
                            return i 
        return -1

# Runtime: 28 ms, faster than 86.09% of Python3 online submissions for Implement strStr().
# Memory Usage: 14.3 MB, less than 77.58% of Python3 online submissions for Implement strStr().
class PrevSolution:
    def strStr(_, haystack: str, needle: str) -> int:
        try:
            return haystack.index(needle)
        except:
            if len(needle) == 0:
                return 0
            else:
                return -1

"""
28. Implement strStr()
Implement strStr().
Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

Clarification:
What should we return when needle is an empty string? This is a great question to ask during an interview.
For the purpose of this problem, we will return 0 when needle is an empty string. 
This is consistent to C's strstr() and Java's indexOf().

Example 1:
Input: haystack = "hello", needle = "ll"
Output: 2

Example 2:
Input: haystack = "aaaaa", needle = "bba"
Output: -1

Example 3:
Input: haystack = "", needle = ""
Output: 0
 
Constraints:
0 <= haystack.length, needle.length <= 5 * 104
haystack and needle consist of only lower-case English characters.
"""

# Testcase 

s = Solution() 
print(s.strStr("hello", "ll")) # 2
print(s.strStr("aaaaa", "bba")) # -1
print(s.strStr("", "a")) # -1
print(s.strStr("", "")) # 0
print(s.strStr("a", "a")) # 0
print(s.strStr("mississippi", "issip")) # 4
print(s.strStr("mississippi", "issipi")) # -1
print(s.strStr("abcd", "cde")) # -1
print(s.strStr("mississippi", "sipp")) # 6
